lib/Dialect/TritonGPU/IR/Dialect.cpp

```diff
@@ -1633,27 +1650,96 @@ LogicalResult PaddedSharedEncodingAttr::verify(
   if (intervalValues.size() != intervals.size())
     return emitError() << "interval values cannot have duplicates";
 
-  if (order.empty())
-    return emitError() << "order cannot be empty";
+  const auto &ll = linearComponent;
+  // The linear layout should mape from [offset, block] to [dim0..dimN). All
+  // bases should be 0 or power of twos and move in a single direction without
+  // broadcasting
+  for (auto outDim : ll.getInDimNames()) {
+    if (outDim.str() != "offset" && outDim.str() != "block")
+      return emitError() << "linearComponent must have `offset` and `block` as "
+                            "only input dims, found: "
+                         << outDim.str();
+  }
+  if (ll.getNumInDims() != 2)
+    return emitError() << "linearComponent must have `offset` and `block` as "
+                          "only input dims";
 
-  if (order.size() != ctaLayout.getRank())
-    return emitError() << "order size (" << order.size()
-                       << ") must match CTALayout rank (" << ctaLayout.getRank()
-                       << ")";
-  return verifyLayoutOrder(emitError, order);
+  // outDims are ['dim0', 'dim1', ...]
+  for (auto [i, dim] : llvm::enumerate(ll.getOutDimNames())) {
+    if (dim.str() != ("dim" + llvm::Twine(i)).str()) {
+      return emitError()
+             << "Expected output dimensions to be ['dim0', 'dim1', ...]. Got "
+             << dim << " at position " << i;
+    }
+  }
+
+  const auto &bases = ll.getBases();
+
+  auto nonZero = [](auto val) { return val != 0; };
+  for (const auto &dimBases : llvm::make_second_range(bases)) {
+    if (!llvm::all_of(dimBases, [&](const auto &basis) {
+          return llvm::count_if(basis, nonZero) <= 1;
+        })) {
+      return emitError()
+             << "Each offset basis must move in at most one dimension.";
+    }
+    if (!llvm::all_of(dimBases, [&](const auto &basis) {
+          return llvm::all_of(
+              basis, [](auto v) { return v == 0 || llvm::isPowerOf2_32(v); });
+        })) {
+      return emitError() << "Each offset basis must be 0 or a power of two.";
+    }
```

Comment:
How are these two any different?

Reply:
Both of them were intended to disallow arbitrary swizzling, and restrict it to reordering of rows and row slices, so we can lower it on GFX9 efficiently (avoid permute). But since this layout is kept general and the restrictions are device specific, I will move those checks to the lowering logic of async_copy/buffer_load_to_local on AMD (follow up pr).

If this makes sense to you?

Reply:
The first `if` checks that we move in a single dimension (one value is non zero) the second `if` checks that all non zero elements are power of twos. This aims to prevent arbitrary swizzling, since we aim at reordering rows/cache line sized chunks of rows.

I added a comment to the second check to make it a bit more clear